### PR TITLE
[SPARK-11720] [SQL][ML] Handle edge cases when count = 0 or 1 for Stats function

### PR DESCRIPTION
**Pull request by JihongMA**
return Double.NaN for mean/average when count == 0 for all numeric types that is converted to Double, Decimal type continue to return null.

### CODE REVIEW

**Comment by vanzin** [comment]
Could you write a PR title that follows the convention used by Spark?
https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark#ContributingtoSpark-PullRequest

**Comment by JihongMA** [comment]
Jenkins, test this please.

**Comment by mengxr** [comment]
LGTM. cc @yhuai for another pass.

**Comment by mengxr** [comment]
@JihongMA Sorry for moving back-and-forth, we need to compromise between SQL behavior and R/Python/others. Basically, we all agree that aggregation should ignore NULL values. So `sum([NULL]) == sum([])` and `sum([NULL, 1.0]) == sum([1.0])`. Then the question is what to return if the collection is empty. In R/Python/Scala, the result is 0.0, but it is NULL in SQL engines. I don't think this is critical because most people won't care an empty collection. Though I prefer returning `0.0`, it is more important to be consistent across Spark SQL. `stddev([1.0])` is different here because we see a value and by definition `stddev = 0.0 / 0 = NaN`.

**Comment by JihongMA** [comment]
@yhuai @mengxr, agreed, it will make the behavior of stats function in consistent across Spark SQL.

**Comment by marmbrus** [comment]
This looks good to me, but I think the title needs to be updated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Kurtosis.scala [hunk 1]
```diff
@@ -37,13 +37,16 @@ case class Kurtosis(child: Expression,
...
     require(moments.length == momentOrder + 1,
       s"$prettyName requires ${momentOrder + 1} central moments, received: ${moments.length}")
     val m2 = moments(2)
     val m4 = moments(4)
 
-    if (n == 0.0 || m2 == 0.0) {
+    if (n == 0.0) {
+      null
+    }
+    else if (m2 == 0.0) {
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Kurtosis.scala]: minor: @JihongMA Could you update your code formatter to put `}` and `else` on the same line?

**Pull request closed without merging**